### PR TITLE
IGNITE-13775 checkpointRWLock wrapper refactoring.

### PR DESCRIPTION
**Pull request by zstan**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/util/ReentrantReadWriteLockWithTracking.java [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+        this.readLockThreshold = readLockThreshold;
+    }
+
+    /** Delegator implementation. */
+    public ReentrantReadWriteLockWithTracking() {
+        readLock = new ReentrantReadWriteLockWithTracking.ReadLock(delegate);
+    }
+
+    /** {@inheritDoc} */
+    @Override public ReentrantReadWriteLockWithTracking.ReadLock readLock() {
```
Comment (akalash) [inline modules/core/src/main/java/org/apache/ignite/internal/util/ReentrantReadWriteLockWithTracking.java:69]: Do you really need ReentrantReadWriteLockWithTracking.ReadLock here? Maybe it makes sense to change it to java.util.concurrent.locks.ReentrantReadWriteLock.ReadLock?

File: modules/core/src/main/java/org/apache/ignite/internal/util/ReentrantReadWriteLockWithTracking.java [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+        return readLock;
+    }
+
+    /** {@inheritDoc} */
+    @Override public ReentrantReadWriteLockWithTracking.WriteLock writeLock() {
```
Comment (akalash) [inline modules/core/src/main/java/org/apache/ignite/internal/util/ReentrantReadWriteLockWithTracking.java:74]: Do you really need ReentrantReadWriteLockWithTracking.WriteLock here? Maybe it makes sense to change it to java.util.concurrent.locks.ReentrantReadWriteLock.WriteLock?

File: modules/core/src/main/java/org/apache/ignite/internal/util/ReentrantReadWriteLockWithTracking.java [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+     * method is designed for use in monitoring system state, not for
+     * synchronization control.
+     * @return the number of read locks held
+     */
+    public int getReadLockCount() {
+        return delegate.getReadLockCount();
+    }
+
+    /** */
+    public static class WriteLock implements Lock {
```
Comment (akalash) [inline modules/core/src/main/java/org/apache/ignite/internal/util/ReentrantReadWriteLockWithTracking.java:116]: Perhaps this class won't make any sense if we change the return type to java.util.concurrent.locks.ReentrantReadWriteLock.WriteLock in the method above.

File: modules/core/src/main/java/org/apache/ignite/internal/util/ReentrantReadWriteLockWithTracking.java [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+            val.set1(cntr);
+
+            READ_LOCK_HOLDER_TS.set(val);
+        }
+
+        /** {@inheritDoc} */
+        @SuppressWarnings("LockAcquiredButNotSafelyReleased")
+        @Override public void lock() {
+            delegate.readLock().lock();
```
Comment (akalash) [inline modules/core/src/main/java/org/apache/ignite/internal/util/ReentrantReadWriteLockWithTracking.java:240]: As I understand here and in other methods super.{method} should be used.

File: modules/core/src/main/java/org/apache/ignite/internal/util/ReentrantReadWriteLockWithTracking.java [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+        /** {@inheritDoc} */
+        @Override public void unlock() {
+            delegate.readLock().unlock();
+
+            dec();
+        }
+    }
+
+    /** Default implementation. */
+    public static class ReadLock implements Lock {
```
Comment (akalash) [inline modules/core/src/main/java/org/apache/ignite/internal/util/ReentrantReadWriteLockWithTracking.java:284]: Can be removed if we will change the return type in the method above.

File: modules/core/src/main/java/org/apache/ignite/internal/util/ReentrantReadWriteLockWithTracking.java (hunk 2 above)
Comment (zstan) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ReentrantReadWriteLockWithTracking.java:74]: I can`t ... cause: protected java.util.concurrent.locks.ReentrantReadWriteLock.WriteLock#WriteLock

**Pull request closed without merging**